File: R/piechart.R

#' Create a pie chart
#'
#' @param data The data frame to consider.
#' @param value The numeric variable to consider.
#' @param label The labeling variable to consider.
#' @param colorCategory A D3 categorical color scheme, you can find more
#' here <https://github.com/d3/d3-scale-chromatic#categorical>. Defaults to 'Paired'
#' @param innerRadius The size of the inner radius of the pie.
#' Defaults to 0. Set the inner radius to a higher value to plot
#' a donut chart.
#' @param outerRadius The size of the outer radius of the pie.
#' @param padRadius From the D3 official documentation, The pad radius compute
#' the fixed linear distance separating adjacent arcs, defined as padRadius * padAngle.
#' @param padAngle Optional. From the D3 official documentation, the padAngle is used to set the
#' padding angle between consecutive arcs.
#' @param cornerRadius From the D3 official documentation, the value of the corner radius for rounded corners.
#' If the corner radius is greater than zero, the corners of the arc are rounded
#' using circles of the given radius. Defaults to 0.
#' @param labelFont The font family of the legend. Defaults to 'sans-serif'.
#' @param title Optional. The title of the plot.
#' @param titleFontSize The font size of the plot title. Defaults to 22.
#' @param font The font family to consider for the titles.
#' Defaults to "Verdana, Geneva, Tahoma, sans-serif".
#' @param bgcol The background color of the SVG. Defaults to "white".
#' @param opacity The color opacity of the pie (from 0 to 1). Defaults to 1.
#' @param labelHeight The height of the legend. Defaults to 18.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.
#'
#' @return A SVG pie chart
#' @export
#'
#' @examples
#' library(dplyr) # needed for the starwars data frame
#'
#' # starwars is part of the dplyr data frame
#' mini_starwars <- starwars %>% tidyr::drop_na(mass) %>%
#'   sample_n(size = 5) # getting 5 random values
#'
#' pieChart(
#'   data = mini_starwars,
#'   value = "mass",
#'   label = "name"
#' )

pieChart <- function(
  data,
  value,
  label,
  colorCategory = "Paired",
  innerRadius = 0,
  outerRadius = "auto",
  padRadius = 0,
  padAngle = NULL,
  cornerRadius = 0,
  labelFont = 'sans-serif',
  title = NULL,
  titleFontSize = 22,
  font = "Verdana, Geneva, Tahoma, sans-serif",
  bgcol = "white",
  opacity = 1,
  labelHeight = 18,
  width = NULL,
  height = NULL
) {

  if (is.null(data[[value]]) || is.null(data[[label]])) {
    stop("Please check that 'value' and 'label' belongs to the specified data frame")
  }


  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }



  r2d3::r2d3(
    data = data,
    script = system.file(
      "d3/scatterplot/piechart.js",
      package = "ddplot"
    ),
    options = list(
      value = value,
      label = label,
      title = title,
      font = font,
      width = width,
      height = height,
      titleFontSize = titleFontSize,
      bgcol = bgcol,
      opacity = opacity,
      colorCategory = colorCategory,
      innerRadius = innerRadius,
      outerRadius = outerRadius,
      labelHeight = labelHeight,
      padRadius = padRadius,
      padAngle = padAngle,
      cornerRadius = cornerRadius,
      labelFont = labelFont
    )
  )
}

File: inst/d3/scatterplot/piechart.js
let margin = 20;

let pie = d3.pie()
    .value(d => d[options.value]);

let radius = Math.min(width, height) / 2 - margin;

if (options.outerRadius != "auto") {
    radius = options.outerRadius
}


let pieData = pie(data);

let colorCategory = "scheme" + options.colorCategory;

let colorSeq = d3.scaleOrdinal(d3[colorCategory]);

let arc = d3.arc()
    .innerRadius(options.innerRadius)
    .outerRadius(radius)
    .padRadius(options.padRadius)
    .padAngle(options.padAngle)
    .cornerRadius(options.cornerRadius)


let labelHeight = options.labelHeight;

svg
.attr('id', options.id)
.append('g')
.attr('transform', `translate(${radius},${radius})`)
.selectAll(null)
    .data(pieData)
    .enter()
    .append('path')
    .attr('d', arc)
    .attr('fill', d => colorSeq(d.data[options.label]))
    .attr('opacity', options.opacity)





const legend = svg
    .append('g')
    .attr('transform', `translate(${radius * 2 + 20},0)`);


legend
    .selectAll(null)
    .data(pieData)
    .enter()
    .append('rect')
    .attr('y', d => 5 + labelHeight * d.index * 1.8)
    .attr('width', labelHeight)
    .attr('height', labelHeight)
    .attr('fill', d => colorSeq(d.data[options.label]))

  legend
    .selectAll(null)
    .data(pieData)
    .enter()
    .append('text')
    .text(d => d.data[options.label])
    .attr('x', labelHeight * 1.2)
    .attr('y', d => labelHeight * d.index * 1.8 + labelHeight)
    .style('font-family', options.labelFont)
    .style('font-size', `${labelHeight}px`);


// Rendering the chart title
svg
  .append("text")
  .attr("x", radius)
  .attr("y", radius * 2 + 30)
  .attr("text-anchor", "middle")
  .attr("dx", "0em")
  .style("font-size", options.titleFontSize)
  .style("font-family", options.font)
  .text(options.title);

